Handle EC2 InvalidInstanceID.NotFound when terminating instances.

scragraham opened the pull request:
Resolves issue #3 

When terminating instances in EC2 it is possible that a subset of instances are already terminated and thus not found by the system. The exception is more likely to originate in the wait_for_status_change() method, but I also handle the possibility of the exception getting thrown when making the call to ec2.terminate_instances.

scragraham committed:
Handle Ec2 terminate_instances better.

 - After terminating instance, while waiting for a status change
   we now account for when an instance id can no longer be found in the cloud.
   - If terminating, we'll remove those instance ids and continue polling
   - If stopping, we'll error out as the instance should still exist
 - Modified wait_for_status_change to pass the list of instance ids
   and a dictionary of EC2 filters instead of querying the instance ids
   and then filtering in python by checking attributes.
 - added helper __terminate_instances() that retries ec2.terminate_instance()
   calls.
 - Modified the methods to also account for the fact that the
   ec2.terminate_instance() method itself could throw the same
   InvalidInstanceID.NotFound error as DescribeInstances.
 - Resolves issue #3

scragraham committed:
adjust retry logic

scragraham committed:
- minor refactor of __terminate_instance() method
 - Added tests of __terminate_instances() and wait_for_status_change
 - minor logging changes to debug instead of info.
 - Added Makefile recipe: test-verbose

scragraham commented:
Unit test results:
```
test_ec2_run_instances (test.test_agent_ec2.TestEC2Agent) ... ok
test_ec2_run_instances_agentruntimeexception (test.test_agent_ec2.TestEC2Agent) ... ok
test_ec2_wait_for_status_change_all_not_found_terminated (test.test_agent_ec2.TestEC2AgentStatusChange) ... ok
test_ec2_wait_for_status_change_already_terminated (test.test_agent_ec2.TestEC2AgentStatusChange) ... ok
test_ec2_wait_for_status_change_invalid_filters (test.test_agent_ec2.TestEC2AgentStatusChange) ... ok
test_ec2_wait_for_status_change_keeps_on_running (test.test_agent_ec2.TestEC2AgentStatusChange) ... ok
test_ec2_wait_for_status_change_multiple_not_found (test.test_agent_ec2.TestEC2AgentStatusChange) ... ok
test_ec2_wait_for_status_change_multiple_not_found_different_order (test.test_agent_ec2.TestEC2AgentStatusChange) ... ok
test_ec2_wait_for_status_change_one_not_found_terminated (test.test_agent_ec2.TestEC2AgentStatusChange) ... ok
test_ec2_wait_for_status_change_stopped_not_found (test.test_agent_ec2.TestEC2AgentStatusChange) ... ok
test_ec2_wait_for_status_change_successfully_terminated (test.test_agent_ec2.TestEC2AgentStatusChange) ... ok
test_ec2_terminate_instances (test.test_agent_ec2.TestEC2AgentTerminateInstances) ... ok
test_ec2_terminate_instances_all_not_found (test.test_agent_ec2.TestEC2AgentTerminateInstances) ... ok
test_ec2_terminate_instances_invalid_id (test.test_agent_ec2.TestEC2AgentTerminateInstances) ... ok
test_ec2_terminate_instances_keeps_running (test.test_agent_ec2.TestEC2AgentTerminateInstances) ... ok
test_ec2_terminate_instances_multiple_invalid_id (test.test_agent_ec2.TestEC2AgentTerminateInstances) ... ok

----------------------------------------------------------------------
Ran 16 tests in 205.215s

OK
```

scragraham committed:
convert set to a list before passing to boto

scragraham commented:
Build: 6389

sjones4 commented on appscale/agents/ec2_agent.py:
```diff
@@ -673,60 +681,124 @@ def terminate_instances(self, parameters):
+          if len(instance_ids) == 0:
+            return True
+
+        # make another attempt
```
For any other ec2 error we may want to have a small delay before retrying and should increment `attempts` to ensure we cannot repeat the same request indefinitely.

scragraham committed:
Adjust retry logic
 - On InvalidInstanceId.NotFound, without counting as an attempt,
   retry if the set of instance_ids is getting smaller and smaller.
 - On other Ec2 errors, wait a bit and retry.
